feature/VIC-665-Registration_Basic_registration_with_a_main_topic_(and_no_PLZ)

patric-dosch-vi opened the pull request.

patric-dosch-vi committed:
feat: VIC-665 Registration | Basic registration with a main topic (and no PLZ)

- add public endpoint to get all active topics (for a tenant)
- the public endpoint do not need auth

patric-dosch-vi commented:
hey @tkuzynow, I think I met everything.
But I have some questions... but I have to write them tomorrow.
I created a draft, if you already want to take a look.
Best, Patric

tkuzynow commented on src/main/java/de/caritas/cob/consultingtypeservice/api/service/TopicServiceFacade.java:
```diff
@@ -24,8 +26,18 @@ public class TopicServiceFacade {
+    return convert(topicEntities);
+  }
+
+  private List<TopicDTO> convert(Collection<TopicEntity> topicEntities) {
```
I would put this convert method to a Converter (to support conversion for collections)

patric-dosch-vi replied:
I thought about that, but it was late ;)
But can introduce this.

tkuzynow replied:
that would be great ;)

tkuzynow commented on src/test/java/de/caritas/cob/consultingtypeservice/api/controller/TopicControllerIT.java:
```diff
@@ -69,7 +70,8 @@ void getTopicList_Should_ReturnTopicsList_When_UserIsAuthenticated()
             get(TopicPathConstants.PATH_GET_TOPIC_LIST)
-                .with(authentication(builder.withAuthority(UserRole.TOPIC_ADMIN.getValue()).build()))
+                .with(
+                    authentication(builder.withAuthority(UserRole.TOPIC_ADMIN.getValue()).build()))
```
get topic list is currently available for just authenticated users, but there is no requirement it has to be the topic admin. I think we need to change it to include again the authentication for the getAll endpoint (@PreAuthorize)

patric-dosch-vi replied:
I think I don't get this. change the test? Or what do you mean?

tkuzynow replied:
no, I mean the code itself, since you introduced publicly available endpoint to get all active topics, we need to enforce that only the topic-admin can access the getAllTopics endpoint (currently it's not the case, it requires authentication, but role is not checked for this endpoint. Please see the TopicController class and look for @PreAuthorize annotations)

patric-dosch-vi replied:
got it

patric-dosch-vi replied:
and the assigning of topics to Beratungsstelle also only the topic-admin can do?

tkuzynow replied:
yes

patric-dosch-vi replied:
if this is still so, I can change this.

tkuzynow reviewed (approved):
I've left two small comments, otherwise the code looks good and can be merged.

patric-dosch-vi commented on src/test/java/de/caritas/cob/consultingtypeservice/api/controller/TopicControllerIT.java:
```diff
@@ -80,6 +82,18 @@ void getTopicList_Should_ReturnTopicsList_When_UserIsAuthenticated()
   }
 
+  @Test
+  void getAllActiveTopics_Should_returnActiveTopicsList() throws Exception {
```
in the Controller-Level we do not have Tenant-Aware-Tests.
How can I introduce this? Because I think this is missing. I added some for the repository, and may also in the Service. But here I am not sure. I think the Tenant Think I still not understood ;)

tkuzynow replied:
you need to create a separate Tenant-Aware class for Test, for example TopicControllerTenantAwareIT. Then please set multitenancy to true with this annotation @TestPropertySource(properties = "multitenancy.enabled=true")

patric-dosch-vi committed:
feat: VIC-665 Registration | Basic registration with a main topic (and no PLZ)

- add converter to convert a Collection of TopicEntities to a List of TopicDTOs
- fix path /topic/public (without ending slash)

patric-dosch-vi committed:
feat: VIC-665 Registration | Basic registration with a main topic (and no PLZ)

- bring NegatedRequestMatcher back
